javatests/arcs/sdk/HandleHolderBaseTest.kt

```diff
@@ -0,0 +1,213 @@
+      assertThat(it.message).isEqualTo(
+        "No dispatcher available for a HandleHolder with no handles."
+      )
+    }
+  }
+
+  @Test
+  fun dispatcher_someHandles_getsFirstHandlesDispatcher() {
+    val handleHolder = DummyHandleHolder()
+    val firstHandle = mock<Handle> {
+      on { dispatcher } doReturn DUMMY_DISPATCHER_1
+    }
+    val secondHandle = mock<Handle> {
+      on { dispatcher } doReturn DUMMY_DISPATCHER_2
+    }
+    handleHolder.setHandle("readHandle", firstHandle)
+    handleHolder.setHandle("writeHandle", secondHandle)
+
+    val actual = handleHolder.dispatcher
+
+    assertThat(actual).isSameInstanceAs(DUMMY_DISPATCHER_1)
+  }
+
+  @Test
+  fun getEntitySpec_handleNameNotInSpec_throwsNoElemException() {
+    val handleHolder = DummyHandleHolder()
+
+    assertFailsWith<NoSuchElementException> {
+      handleHolder.getEntitySpecs("notInSpec")
+    }.also {
+      assertThat(it.message).isEqualTo(
+        "Particle TestParticle does not have a handle with name notInSpec."
+      )
+    }
+  }
+
+  @Test
+  fun getEntitySpec_handleNameInSpec_returnsSpec() {
+    val handleHolder = DummyHandleHolder()
+
+    assertThat(handleHolder.getEntitySpecs("readHandle")).isNotEmpty()
+  }
+
+  @Test
+  fun getHandle_handleNameNotInSpec_throwsNoElemException() {
+    val handleHolder = DummyHandleHolder()
+    handleHolder.setHandle("readHandle", mock())
+
+    assertFailsWith<IllegalArgumentException> {
+      handleHolder.setHandle("readHandle", mock())
+    }.also {
+      assertThat(it.message).isEqualTo("TestParticle.readHandle has already been initialized.")
+    }
+  }
+
+  @Test
+  fun getHandle_handleNameInSpec_returnsHandle() {
+    val handleHolder = DummyHandleHolder()
+    val mock = mock<Handle>()
+    handleHolder.setHandle("writeHandle", mock)
+
+    assertThat(handleHolder.getHandle("writeHandle")).isSameInstanceAs(mock)
+  }
+
+  @Test
+  fun setHandle_handelAlreadySet_asserts() {
+    val handleHolder = DummyHandleHolder()
+    handleHolder.setHandle("writeHandle", mock())
+
+    assertFailsWith<IllegalArgumentException> {
+      handleHolder.setHandle("writeHandle", mock())
+    }.also {
+      assertThat(it.message).isEqualTo(
+        "TestParticle.writeHandle has already been initialized."
+      )
+    }
+  }
+
+  @Test
+  fun setHandle_handelNameNotInSpec_throwsNoElemException() {
+    val handleHolder = DummyHandleHolder()
+
+    assertFailsWith<NoSuchElementException> {
+      handleHolder.setHandle("handleButNotEntities", mock())
+    }.also {
+      assertThat(it.message).isEqualTo(
+        "Particle TestParticle does not have a handle with name handleButNotEntities."
+      )
+    }
+  }
+
+  @Test
+  fun setHandle_handelNameInSpec_success() {
+    val handleHolder = DummyHandleHolder()
+    val mock = mock<Handle>()
+
+    handleHolder.setHandle("writeHandle", mock)
+
+    assertThat(handleHolder.writeHandle).isSameInstanceAs(mock)
+  }
+
+  @Test
+  fun isEmpty_delegatesToHandlesMap() {
```

Comment:
This test name feels a bit inaccurate, and could be broken into two tests:

`isEmpty_withNoHandlesSet_returnsFalse`

`isEmpty_withHandlesSet_returnsTrue`

Reply:
(maybe inaccurate is not the right word... what I mean to say is that it's not really explaining the behavior that we care about, but rather, an internal implementation detail)